### PR TITLE
[cli] Add multisig:approve 

### PR DESCRIPTION
**Pull request by timmoreton**
### Description

Add `multisig:approve` to CLI.

```
bin/run multisig:approve --help
Approves an existing transaction on a multi-sig contract

USAGE
  $ celocli multisig:approve

OPTIONS
  --for=0xc1912fEE45d61C87Cc5EA59DaE31190FFFFf232d   (required) Account Address
  --from=0xc1912fEE45d61C87Cc5EA59DaE31190FFFFf232d  (required) Account Address
  --globalHelp                                       View all available global flags
  --tx=tx                                            (required) Transaction to approve

EXAMPLE
  approve --from 0x6ecbe1db9ef729cbe972c83fb886247691fb6beb --for 0x5409ed021d9299bf6814279a6a1411a7e866a631 --tx
  3
```

### Other changes

`celocli multisig:show` was not showing calldata correctly unless the multisig instance specifically had a sourcify mapping on Explorer. Now it uses the mapping for the Multisig.sol contract automatically.

### Tested

By creating a multisig on alfajores, generating a pending tx, viewing it, then using the changes to approve the tx. 

### Backwards compatibility

Yes

### Documentation

To follow

### CODE REVIEW

**Comment by arthurgousset** [comment]
(for future reference)
@nicolasbrugneaux and I agreed to pair on reviewing/testing this PR tomorrow. 

In general, the objective is that I get more comfortable with the CLI codebase.

**Comment by 0xarthurxyz** [comment]
Noting this is for governance approver changes.

### `celocli multisig:approve`

I'm struggling to test this with an example without having a multisig to test with. There doesn't seem to be a celocli command to create a multisig only:

```
multisig:approve   Approves an existing transaction on a multi-sig contract
multisig:show      Shows information about multi-sig contract
multisig:transfer  Ability to approve CELO transfers to and from multisig. Submit transaction or approve a matching existing transaction
````

### `celocli multisig:show`

I'm struggling to test this with an example without having a multisig to test with.

File: packages/cli/src/commands/multisig/approve.ts
Comment (arthurgousset) [inline packages/cli/src/commands/multisig/approve.ts:13]: This would benefit from a description. Maybe something like this? 

```suggestion
      required: true,
      description: 'Account approving the multi-sig transaction',
```

File: packages/cli/src/commands/multisig/approve.ts
Comment (arthurgousset) [inline packages/cli/src/commands/multisig/approve.ts:17]: This would benefit from a description. Maybe something like this? 

```suggestion
      required: true,
      description: 'Address of the multi-sig contract',
```

File: packages/cli/src/commands/multisig/approve.ts
Comment (arthurgousset) [inline packages/cli/src/commands/multisig/approve.ts:51]: ### Note-to-self (please ignore)

The method signature of `displaySendTx` is a little cryptic for a newcomer (like me). Could be "nice to have" refactor some day.

```ts
displaySendTx<string | void | boolean>('approveTx', tx, {}, 'ApproveTx')
```

File: packages/cli/src/commands/multisig/show.ts
Comment (arthurgousset) [inline packages/cli/src/commands/multisig/show.ts:41]: ### Notes-to-self (please ignore)

As Tim mentioned, this additional line ensures that, given a valid multi-sig instance (`args.address`), its relevant contract details are available (importantly its ABI), by fetching it from the `Multisig.sol` core contract (`CeloContract.MultiSig`).

The ABI is then used to correctly parse transaction data in 

```ts
tryParseTxInput(txdata.destination, txdata.data)
```

a couple lines below.

For future referece, [`updateContractDetailsMapping`](https://github.com/celo-org/celo-monorepo/blob/95736497e4f2e15f3247357b7776e6ee2440f7c4/packages/sdk/explorer/src/block-explorer.ts#L76) is defined in `@celo/explorer` and uses the following helpers [`getContractDetailsFromContract`](https://github.com/celo-org/celo-monorepo/blob/95736497e4f2e15f3247357b7776e6ee2440f7c4/packages/sdk/explorer/src/base.ts#L17-L30) and [`getContractMappingFromDetails`](https://github.com/celo-org/celo-monorepo/blob/95736497e4f2e15f3247357b7776e6ee2440f7c4/packages/sdk/explorer/src/block-explorer.ts#L53)

File: packages/docs/command-line-interface/multisig.md
Comment (arthurgousset) [inline packages/docs/command-line-interface/multisig.md]: Reminder: Note to ensure these two descriptions are automagically updated when the suggested description for `from` and `for` is committed in [`approve.ts`](https://github.com/celo-org/celo-monorepo/blob/95736497e4f2e15f3247357b7776e6ee2440f7c4/packages/cli/src/commands/multisig/approve.ts#L12-L17).

File: packages/sdk/contractkit/src/wrappers/MultiSig.ts
Comment (arthurgousset) [inline packages/sdk/contractkit/src/wrappers/MultiSig.ts:68]: Nit: Any reason you didn't use the existing [`submitOrConfirmTransaction`](https://github.com/celo-org/celo-monorepo/blob/95736497e4f2e15f3247357b7776e6ee2440f7c4/packages/sdk/contractkit/src/wrappers/MultiSig.ts#L51-L54) function and instead created the new `confirmTransaction` function in ContractKit above?

I can see it makes the function call shorter in [`approve.ts`](https://github.com/celo-org/celo-monorepo/blob/1a0a5b2b1d1a8b97f0e36ac7266d071be4b7a0b3/packages/cli/src/commands/multisig/approve.ts#L44), but in exchange adds complexity to ContractKit's API. 

User's now have a choice between `submitOrConfirmTransaction` and `confirmTransaction`.

I'm not strongly for or against either approach (I haven't spent enough time in CK and the CLI), just curious if there was a reason you decided to add a function.

**Review by arthurgousset — COMMENTED** [review]
Thanks for this PR @timmoreton ! 

@nicolasbrugneaux and I paired on this, and I left a couple of comments. 

The implementation looks good, the only thing is we couldn't test `celocli multisig:approve` without an example multisig to test with.

We couldn't find a command to create a multisig with via the celocli, the only commands are:

```sh
multisig:approve   Approves an existing transaction on a multi-sig contract
multisig:show      Shows information about multi-sig contract
multisig:transfer  Ability to approve CELO transfers to and from multisig. Submit transaction or approve a matching existing transaction
````

Presumably I'll need to create a multi-sig instance using ContractKit and a Node.js script.

File: packages/sdk/contractkit/src/wrappers/MultiSig.ts
Comment (timmoreton) [reply packages/sdk/contractkit/src/wrappers/MultiSig.ts:68]: `submitOrConfirmTransaction` will search for and approve if present or else submit a tx matching given bytes -- here the whole idea is the tx already exists, we know it's id, and we definitely don't want to ever submit a new one. So more efficient to expose `confirmTransaction` directly.

**Comment by timmoreton** [comment]
> Thanks for this PR @timmoreton !
> 
> @nicolasbrugneaux and I paired on this, and I left a couple of comments.
> 
> The implementation looks good, the only thing is we couldn't test `celocli multisig:approve` without an example multisig to test with.
> 
> We couldn't find a command to create a multisig with via the celocli, the only commands are:
> 
> ```shell
> multisig:approve   Approves an existing transaction on a multi-sig contract
> multisig:show      Shows information about multi-sig contract
> multisig:transfer  Ability to approve CELO transfers to and from multisig. Submit transaction or approve a matching existing transaction
> ```
> 
> Presumably I'll need to create a multi-sig instance using ContractKit and a Node.js script.

Celo Terminal can create multisigs -- start it with Alfajores like this:

```
export CELOTERMINAL_NETWORK_ID=44787
export CELOTERMINAL_ACCOUNTS_DB=home/.celoterminal/celoaccounts-test.db
/Applications/Celo\ Terminal.app/Contents/MacOS/Celo\ Terminal
```

You can look at the transactions and logs around [this multisig I created](https://alfajores.celoscan.io/address/0x05a37343c58a3af320b181c3d365b35613570878) to see how I tested it

**Comment by 0xarthurxyz** [comment]
Thanks @timmoreton 💯 

(_for future reference_) 

## Steps to test `multisig` CLI command on Alfajores

### Create a MultiSig on Alfajores using Celo Terminal

1. Download [Celo Terminal](https://celoterminal.com/)
2. Run this in your CLI (this will start Celo Terminal on Alfajores, instead of Mainnet)

```sh
export CELOTERMINAL_NETWORK_ID=44787
export CELOTERMINAL_ACCOUNTS_DB=home/.celoterminal/celoaccounts-test.db
/Applications/Celo\ Terminal.app/Contents/MacOS/Celo\ Terminal
```

3. Create a Multisig ("Add account" > "Create a MultiSig account", you will first need a standard account, since the MultiSig needs an owner)

image

**Review by arthurgousset — APPROVED** [review]
Thanks @timmoreton! 

Tested the `multisig:approve` command with a MultiSig account on Alfajores. 

Works as expected ✅

I left two improvements for a later time and documented the steps to test the CLI with a MultiSig account on Alfajores. We can merge and release this for now.

## Future improvements

### Improve descriptions to emphasize zero indexed transactions

I first ran the command with `--tx 1` instead of `--tx 0`, because that was the number of transactions shown in the `multisig:show` output. But transactions are zero indexed, so if there is 1 transaction, then the transaction to approve is transaction index 0 (`--tx 0`). 

### Improve error message when tx index is larger than tx count

Running the command with a transaction index higher than the number of transactions (e.g. `--tx 1` instead of `--tx 0`) threw an error `1 is existing transaction`, which is incorrect since transaction index 1 is not an existing transaction. It should probably say "is not an existing transaction".

```sh
bin/run multisig:approve --from 0xEb08a36D153827CC750dd264c9e7360A3174Dc27 --for 0x37A6bA9CFC3C4C7f82bDf1Fe1B7cEb410E78738E > --tx 1 --privateKey <privateKey>

Running Checks:
  ✔  The provided address is an owner of the multisig 
  ✘  1 is existing transaction 
›   Error: Some checks didn't pass!
```

## Steps to test `celocli multisig` on Alfajores

### Create a MultiSig on Alfajores using Celo Terminal

1. Download [Celo Terminal](https://celoterminal.com/)
2. Run this in your CLI (this will start Celo Terminal on Alfajores, instead of Mainnet)

  ```sh
  export CELOTERMINAL_NETWORK_ID=44787
  export CELOTERMINAL_ACCOUNTS_DB=home/.celoterminal/celoaccounts-test.db
  /Applications/Celo\ Terminal.app/Contents/MacOS/Celo\ Terminal
  ```

3. Add 2 "standard accounts" (the multisig will need two owners for testing)
4. Create a Multisig ("Add account" > "Create a MultiSig account", set both accounts as owners)

image

5. Faucet your new MultiSig ([faucet.celo.org](https://faucet.celo.org/))

### Make a test transfer from the MultiSig

6. Make a test transfer from the MultiSig (to anywhere, e.g. one of your standard accounts)

image

### Approve the MultiSig transfer from the CLI

7. Checkout branch

```sh
git checkout timmoreton/multisig-approve-cli
```

8. [Build branch](https://github.com/celo-org/celo-monorepo/blob/master/SETUP.md#building-celo-monorepo )

```sh
yarn 
yarn build --ignore docs
```

9. Navigate to locally built CLI package

```sh
cd packages/cli
```

10. [Set RPC node and network to Alfajores](https://docs.celo.org/cli#optional-run-a-full-node ) (for testing)

```sh
bin/run config:set https://alfajores-forno.celo-testnet.org
```

11. Check RPC node and network is set

```sh
bin/run config:get                           

node: https://alfajores-forno.celo-testnet.org
gasCurrency: auto
```

12. Confirm the MultiSig transaction is queued (I'm using my test MultiSig created above)

```sh
bin/run multisig:show 0x37A6bA9CFC3C4C7f82bDf1Fe1B7cEb410E78738E

Owners: 
  0: 0x303C22e6ef01CbA9d03259248863836CB91336D5
  1: 0xEb08a36D153827CC750dd264c9e7360A3174Dc27
Required confirmations: 2 
Required confirmations (internal): 2 
Transactions: 1
```

13. Approve the transaction (I removed my private key from the command for obvious reasons)

```sh
bin/run multisig:approve --from 0xEb08a36D153827CC750dd264c9e7360A3174Dc27 --for 0x37A6bA9CFC3C4C7f82bDf1Fe1B7cEb410E78738E --tx 0 --privateKey <privateKey>

Running Checks:
   ✔  The provided address is an owner of the multisig 
   ✔  0 is existing transaction 
All checks passed
SendTransaction: approveTx
txHash: 0xfe52b86eff92a715ffe158e0ea6d94274e5e083dcf2f40e8baca6a2c638b655e
Sending Transaction: approveTx... done
```

14. Confirm Multisig transfer worked using transaction hash above: https://alfajores.celoscan.io/tx/0xfe52b86eff92a715ffe158e0ea6d94274e5e083dcf2f40e8baca6a2c638b655e 

✅ Yes, MultiSig balance changed 5 CELO -> 4.5, and 0.5 CELO was transferred from MultiSig to chosen recipient.

File: packages/cli/src/commands/multisig/approve.ts
Comment (arthurgousset) [inline packages/cli/src/commands/multisig/approve.ts]: While we're waiting for primitive tests to pass, here is one of the small improvements I suggested. 

There is 1 pending transaction:

```sh
$ bin/run multisig:show 0x37A6bA9CFC3C4C7f82bDf1Fe1B7cEb410E78738E

Owners: 
  0: 0x303C22e6ef01CbA9d03259248863836CB91336D5
  1: 0xEb08a36D153827CC750dd264c9e7360A3174Dc27
Required confirmations: 2 
Required confirmations (internal): 2 
Transactions: 1
```

So using `--tx 1` would be incorrect, because transactions are zero indexed, instead `--tx 0` is correct. But, the error message incorrectly states `✘  1 is existing transaction`, when it should probably say `✘  1 is not an existing transaction`.

```sh
$ bin/run multisig:approve --from 0xEb08a36D153827CC750dd264c9e7360A3174Dc27 --for 0x37A6bA9CFC3C4C7f82bDf1Fe1B7cEb410E78738E > --tx 1 --privateKey <privateKey>

Running Checks:
  ✔  The provided address is an owner of the multisig 
  ✘  1 is existing transaction 
›   Error: Some checks didn't pass!
```

```suggestion
      .addCheck(`${res.flags.tx} is not an existing transaction`, async () => {
```

**Review by arthurgousset — CHANGES_REQUESTED** [review]
Added a small improvement while we're waiting for primitive tests to pass.

File: packages/cli/src/commands/multisig/approve.ts
Comment (aaronmgdr) [reply packages/cli/src/commands/multisig/approve.ts]: this would be wrong. because we WANT it to be an existing transaction. 

do what you want "show WHAT the error is when it failes" we will need to rewrite how addCheck works to take both a happy and sad messege

Comment (arthurgousset) [reply packages/cli/src/commands/multisig/approve.ts]: Ah, I see 👍 Looks like `addCheck` takes an optional error message:

```ts
addCheck(name: string, predicate: () => boolean | Promise<boolean>, errorMessage?: string | undefined)
```

I tried giving it an error message here for example: 

```diff
const checkBuilder = newCheckBuilder(this)
      .isMultiSigOwner(account, multisig)
      .addCheck(
        `${res.flags.tx} is existing transaction`,
        async () => {
          const max = await multisig.getTransactionCount(true, true)
          return res.flags.tx < max
        },
+        `${res.flags.tx} is not an existing transaction`
      )
```

But the CLI output is suboptimal: 

```sh
bin/run multisig:approve --from 0xEb08a36D153827CC750dd264c9e7360A3174Dc27 --for 0x37A6bA9CFC3C4C7f82bDf1Fe1B7cEb410E78738E --tx 10 --privateKey <privatekey>

Running Checks:
   ✔  The provided address is an owner of the multisig 
   ✘  10 is existing transaction 10 is not an existing transaction
 ›   Error: Some checks didn't pass!
```

Let me re-write these a little to make this work.

Comment (arthurgousset) [reply packages/cli/src/commands/multisig/approve.ts]: How about this? 

```diff
     const checkBuilder = newCheckBuilder(this)
       .isMultiSigOwner(account, multisig)
-      .addCheck(`${res.flags.tx} is existing transaction`, async () => {
-        const max = await multisig.getTransactionCount(true, true)
-        return res.flags.tx < max
-      })
+      .addCheck(
+        `Checking that ${res.flags.tx} is an existing transaction.`,
+        async () => {
+          const max = await multisig.getTransactionCount(true, true)
+          return res.flags.tx < max
+        },
+        `${res.flags.tx} is not an existing transaction.`
+      )
```

Gives the following output? 

Error: 

```sh
bin/run multisig:approve --from 0xEb08a36D153827CC750dd264c9e7360A3174Dc27 --for 0x37A6bA9CFC3C4C7f82bDf1Fe1B7cEb410E78738E --tx 10 --privateKey <privateKey>

Running Checks:
   ✔  The provided address is an owner of the multisig 
   ✘  Checking that 10 is an existing transaction. 10 is not an existing transaction.
 ›   Error: Some checks didn't pass!
```

Success:

```sh
bin/run multisig:approve --from 0xEb08a36D153827CC750dd264c9e7360A3174Dc27 --for 0x37A6bA9CFC3C4C7f82bDf1Fe1B7cEb410E78738E --tx 0 --privateKey <privateKey>

Running Checks:
   ✔  The provided address is an owner of the multisig 
   ✔  Checking that 0 is an existing transaction. 
All checks passed
```

Comment (aaronmgdr) [reply packages/cli/src/commands/multisig/approve.ts]: `Checking that 10 is an existing transaction. (No transaction with ID 10)`

`Checking that 10 is an existing transaction. (Failed: no transaction found)`

`Checking that 10 is an existing transaction. (It is not)`

`Checking that 10 is an existing transaction. But could not find it`

Comment (aaronmgdr) [reply packages/cli/src/commands/multisig/approve.ts]: some ideas if you like

Comment (arthurgousset) [reply packages/cli/src/commands/multisig/approve.ts]: Nice, good suggestions. I chose the below.

Error:

```sh
Running Checks:
   ✔  The provided address is an owner of the multisig 
   ✘  Checking that 10 is an existing transaction. (Failed: No transaction with index 10 found)
 ›   Error: Some checks didn't pass!
```

Success:

```sh
Running Checks:
   ✔  The provided address is an owner of the multisig 
   ✔  Checking that 0 is an existing transaction. 
All checks passed
```

If you're happy with that, I'll make a small commit, run the tests again (😿), and merge it, so this can be released with the other CLI changes: 

- https://github.com/celo-org/celo-monorepo/pull/10699

Comment (arthurgousset) [reply packages/cli/src/commands/multisig/approve.ts]: Added in this commit https://github.com/celo-org/celo-monorepo/pull/10693/commits/54c0feca773442b44a65c03e2742f675ea0ed70a

**Review by aaronmgdr — APPROVED** [review]

**Review by arthurgousset — APPROVED** [review]

**Pull request merged**
